server/remote_cache/byte_stream_server/byte_stream_server.go

```diff
@@ -152,11 +162,17 @@ func (s *ByteStreamServer) Read(req *bspb.ReadRequest, stream bspb.ByteStream_Re
 // `complete` or not.
 
 type writeState struct {
-	writer             io.WriteCloser
-	checksum           hash.Hash
+	// Top-level writer that handles incoming bytes.
+	writer io.Writer
+	// Optional writer that decompresses incoming bytes and writes them to both
+	// the checksum and the cache.
+	decompressor io.WriteCloser
+	// Writer which writes uncompressed bytes to the cache.
+	cacheWriter        io.WriteCloser
```

Comment:
this is confusing -- we went from one WriteCloser to 3 (well, 2 WriteClosers and 1 Writer).

I think the complexity is happening because we're using io.MultiWriter -- could we add a MultiWriteCloser and stick with the one WriteCloser?

hint: https://gist.github.com/wader/d82787b91391b71cf39d4f55d98c6d88